### PR TITLE
fix issues with pushing, improve logging and metrics

### PR DESCRIPTION
**Pull request by rajatjindal**
fix issues with pushing, improve logging and metrics

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -202,9 +176,27 @@ func ParseCfgFile(cfgFile string) (*Config, error) {
...
+	err = validateMetricRelabelConfigs(cfg.MetricRelabelConfigs)
+	if err != nil {
+		fmt.Println(err)
+		return nil, err
+	}
+
 	return cfg, nil
 }
 
+func validateMetricRelabelConfigs(metricRelabelConfigs []*RelabelConfig) error {
```
Comment (johngmyers) [inline config/config.go]: Shouldn't there be an added test case covering this?

File: example/service.yaml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    service.beta.kubernetes.io/aws-load-balancer-backend-protocol: http
+    service.beta.kubernetes.io/aws-load-balancer-internal: 0.0.0.0/0
+    service.beta.kubernetes.io/aws-load-balancer-ssl-ports: http
+    service.beta.kubernetes.io/aws-load-balancer-proxy-protocol: '*'
+    service.beta.kubernetes.io/aws-load-balancer-ssl-cert: <CERT-ARN>
+spec:
+  ports:
+  - port: 443
+    targetPort: 9201
+    name: http
```
Comment (johngmyers) [inline example/service.yaml]: Shouldn't the name be "https"?

File: config/config.go [hunk 3]
```diff
@@ -132,31 +134,30 @@ const (
 )
 
 // ParseCfgFile read the provided config file and parse it into config object
-func ParseCfgFile(cfgFile string) (*Config, error) {
+func ParseCfgFile(cfgFile string, cfg *Config) error {
```
Comment (johngmyers) [inline config/config.go]: Why this change? If the kairosdb-url is missing, the new code will give the caller an inconsistent Config along with their error.

File: kairosdb/client.go [hunk 4]
```diff
@@ -86,6 +93,9 @@ func (c *Client) Send(samples model.Samples) (err error) {
 	begin := time.Now()
 
 	err = c.write(datapoints)
+	if err != nil {
+		logrus.Errorf("failed writing metrics to downstream. error: %s", err)
```
Comment (johngmyers) [inline kairosdb/client.go]: Don't you need to increment a counter here?

File: kairosdb/client.go [hunk 5]
```diff
@@ -113,24 +123,36 @@ func (c *Client) write(datapoints []*DataPoint) error {
...
+		failedSamples.WithLabelValues(c.name()).Add(float64(totalRequests))
 		return err
 	}
+
 	defer resp.Body.Close()
 
+	if resp == nil {
+		failedSamples.WithLabelValues(c.name()).Add(float64(totalRequests))
+		logrus.Errorf("no response received")
+		return fmt.Errorf("no response received")
```
Comment (johngmyers) [inline kairosdb/client.go]: Shouldn't you either log or return an error, but not both?

File: kairosdb/client.go [hunk 6]
```diff
@@ -113,24 +123,36 @@ func (c *Client) write(datapoints []*DataPoint) error {
...
 	if resp.StatusCode == http.StatusNoContent {
+		sentSamples.WithLabelValues(c.name()).Add(float64(totalRequests))
 		return nil
 	}
 
 	// API returns status code 400 on error, encoding error details in the
 	// response content in JSON.
-	buf, err = ioutil.ReadAll(resp.Body)
 
 	if err != nil {
```
Comment (johngmyers) [inline kairosdb/client.go]: Since you removed the line above, this can never be true.

File: kairosdb/client.go [hunk 7]
```diff
@@ -140,7 +162,7 @@ func (c *Client) write(datapoints []*DataPoint) error {
 	sentSamples.WithLabelValues(c.name()).Add(float64(successful))
 	failedSamples.WithLabelValues(c.name()).Add(float64(failed))
 
-	return fmt.Errorf("failed to write %d samples to KairosDB, %d succeeded", len(r["errors"]), totalRequests-len(r["errors"]))
+	return fmt.Errorf("failed to write [%d] samples from [%d]", failed, totalRequests)
```
Comment (johngmyers) [inline kairosdb/client.go]: "of" instead of "from".

File: kairosdb/client.go [hunk 8]
```diff
@@ -31,6 +30,13 @@ var (
 		},
 		[]string{"remote"},
 	)
+	unknownStatusSamples = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "unknown_status_samples_total",
+			Help: "Total number of samples which we can't tell if they were pushed downstream successfully.",
```
Comment (johngmyers) [inline kairosdb/client.go]: Perhaps "Total number of samples sent without receiving a response."

File: kairosdb/client.go [hunk 9]
```diff
@@ -136,21 +137,23 @@ func (c *Client) write(datapoints []*DataPoint) error {
...
 		logrus.Errorf("%s", err)
 		unknownStatusSamples.WithLabelValues(c.name()).Add(float64(totalRequests))
 		return err
 	}
 
 	var r map[string][]interface{}
-	if err = json.Unmarshal(buf, &r); err != nil {
+	if err = json.Unmarshal(respbuf, &r); err != nil {
+		logrus.Errorf("response received is : %s", string(respbuf))
 		logrus.Errorf("%s", err)
```
Comment (johngmyers) [inline kairosdb/client.go]: Perhaps combine into a single log line?

File: kairosdb/client.go [hunk 10]
```diff
@@ -159,6 +162,14 @@ func (c *Client) write(datapoints []*DataPoint) error {
 	failed := len(r["errors"])
 	successful := totalRequests - failed
 
+	//unlikely, but will keep it here anyways
+	//added because a code issue was causing this condition
+	if successful < 0 {
+		logrus.Errorf("response from kairosdb %v", r)
+		logrus.Errorf("req to kairosdb %v", string(buf))
+		return fmt.Errorf("no of failed datapoints [%d] is greater than total datapoints [%d]", failed, totalRequests)
```
Comment (johngmyers) [inline kairosdb/client.go]: You can spell out "number" rather than abbreviating it.

File: example/configmap.yaml [hunk 11]
```diff
@@ -4,7 +4,7 @@ metadata:
   name: prom-to-kairosdb
 data:
   config.yaml: |
-    kairosdb-url: "http://localhost/1234"
+    kairosdb-url: "http://kairosdb-api-url.yourdomain.com"
```
Comment (johngmyers) [inline example/configmap.yaml]: Use example.com

File: kairosdb/client.go (hunk 4 above)
Comment (rajatjindal) [reply kairosdb/client.go]: the metrics are incremented in the write function. so not needed here

File: kairosdb/client.go (hunk 5 above)
Comment (rajatjindal) [reply kairosdb/client.go]: sure, i think we need to return error

File: example/configmap.yaml (hunk 11 above)
Comment (rajatjindal) [reply example/configmap.yaml]: done

File: kairosdb/client.go (hunk 10 above)
Comment (rajatjindal) [reply kairosdb/client.go]: done

File: kairosdb/client.go (hunk 8 above)
Comment (rajatjindal) [reply kairosdb/client.go]: done

File: kairosdb/client.go (hunk 7 above)
Comment (rajatjindal) [reply kairosdb/client.go]: done

File: kairosdb/client.go (hunk 5 above)
Comment (rajatjindal) [reply kairosdb/client.go]: done

File: example/service.yaml (hunk 2 above)
Comment (rajatjindal) [reply example/service.yaml]: done, also changed the annotation

File: config/config.go (hunk 3 above)
Comment (rajatjindal) [reply config/config.go]: done. reverted this change.

File: config/config.go (hunk 1 above)
Comment (rajatjindal) [reply config/config.go]: added few more test cases.

**Comment by rajatjindal** [comment]
@johngmyers can you please re-review this PR

File: config/config.go [hunk 12]
```diff
@@ -192,21 +170,64 @@ func ParseCfgFile(cfgFile string) (*Config, error) {
...
 			Prefix:       cfg.MetricnamePrefix,
 		}
 
 		cfg.MetricRelabelConfigs = append(cfg.MetricRelabelConfigs, relabelConfig)
 	}
 
+	err = validateMetricRelabelConfigs(cfg.MetricRelabelConfigs)
+	if err != nil {
+		logrus.Errorf("%s", err)
+		return nil, err
```
Comment (johngmyers) [inline config/config.go]: Probably shouldn't both log and return.

File: config/config.go [hunk 13]
```diff
@@ -224,29 +245,29 @@ func getAbsFilename(cfgFile string) (string, error) {
...
+	return filepath.Dir(ex), nil
 }
 
 func ValidateFile(cfgFile string) (bool, error) {
 	fstat, err := os.Stat(cfgFile)
 	if os.IsNotExist(err) {
-		fmt.Printf("%s dont exist\n", cfgFile)
-		return false, fmt.Errorf("file not found\n")
+		logrus.Errorf("%s dont exist\n", cfgFile)
+		return false, fmt.Errorf("file not found")
```
Comment (johngmyers) [inline config/config.go]: Probably shouldn't both log and return error.

File: kairosdb/client.go [hunk 14]
```diff
@@ -95,45 +107,72 @@ func (c *Client) Send(samples model.Samples) (err error) {
...
 	}
 
 	// API returns status code 400 on error, encoding error details in the
 	// response content in JSON.
-	buf, err = ioutil.ReadAll(resp.Body)
-
+	respbuf, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
+		logrus.Errorf("%s", err)
+		unknownStatusSamples.WithLabelValues(c.name()).Add(float64(totalRequests))
```
Comment (johngmyers) [inline kairosdb/client.go]: Probably shouldn't both log and return error.
